src/Controls/src/Core/View/View.cs

```diff
@@ -118,41 +119,46 @@ void RemoveItems(IEnumerable<IElementDefinition> elements)
 				switch (args.Action)
 				{
 					case NotifyCollectionChangedAction.Add:
-						AddItems(args.NewItems.OfType<IElementDefinition>());
+						AddItems(args.NewItems);
 						break;
 					case NotifyCollectionChangedAction.Remove:
-						RemoveItems(args.OldItems.OfType<IElementDefinition>());
+						RemoveItems(args.OldItems);
 						break;
 					case NotifyCollectionChangedAction.Replace:
-						AddItems(args.NewItems.OfType<IElementDefinition>());
-						RemoveItems(args.OldItems.OfType<IElementDefinition>());
+						AddItems(args.NewItems);
+						RemoveItems(args.OldItems);
 						break;
 					case NotifyCollectionChangedAction.Reset:
 
-						List<IElementDefinition> remove = new List<IElementDefinition>();
-						List<IElementDefinition> add = new List<IElementDefinition>();
-
-						foreach (IElementDefinition item in _gestureRecognizers.OfType<IElementDefinition>())
+						foreach (IGestureRecognizer gestureRecognizer in _gestureRecognizers)
 						{
-							if (!_gestureRecognizers.Contains((IGestureRecognizer)item))
-								add.Add(item);
-							item.Parent = this;
+							if (gestureRecognizer is IElementDefinition item)
+							{
+								item.Parent = this;
+							}
 						}
 
-						foreach (IElementDefinition item in GestureController.CompositeGestureRecognizers.OfType<IElementDefinition>())
-						{
-							if (item == _recognizerForPointerOverState)
-								continue;
+						var compositeGestureRecognizers = GestureController.CompositeGestureRecognizers.ToHashSet();
 
-							if (_gestureRecognizers.Contains((IGestureRecognizer)item))
-								item.Parent = this;
-							else
-								remove.Add(item);
+						foreach (IGestureRecognizer gestureRecognizer in compositeGestureRecognizers)
```

Comment:
I likely have to copy CompositeGestureRecognizers to some kind of list, because otherwise it would result in the dreaded `collection was modified` error. I don't think that can be avoided.

A for loop can sometimes be a tiny bit faster, but I didn't see a difference.

Reply:
Actually you are right about a for loop removing the need for copying the list, but running the benchmark without a HashSet we can see that the ClearLotsOfGestureRecognizers benchmark is slower by more than an order of magnitude.


| Method                         | Mean         | Error      | StdDev     | Gen0      | Gen1     | Gen2    | Allocated  |
|------------------------------- |-------------:|-----------:|-----------:|----------:|---------:|--------:|-----------:|
| AddLotsOfGestureRecognizers    | 10,849.69 us | 262.116 us | 717.538 us | 1000.0000 | 125.0000 | 31.2500 | 9154.29 KB |
| ClearLotsOfGestureRecognizers  |  6,903.05 us | 134.973 us | 119.650 us |  164.0625 | 148.4375 | 54.6875 | 1551.17 KB |
| RemoveLotsOfGestureRecognizers |     52.28 us |   1.040 us |   1.068 us |    0.3052 |   0.2441 |       - |    2.93 KB |
| AddOneGestureRecognizer        |  3,554.05 us |  24.573 us |  19.185 us |  433.5938 | 429.6875 | 15.6250 | 3992.92 KB |
| RemoveOneGestureRecognizer     |  2,316.22 us |  44.572 us |  49.542 us |  292.9688 | 289.0625 |       - | 2719.46 KB |
| ClearOneGestureRecognizer      |  3,520.75 us |  59.967 us |  56.093 us |  523.4375 | 519.5313 |  3.9063 | 4812.63 KB |